File: Pre CAT/DS/QuickSort.c
#include<stdio.h>
int partition(int *arr,int left,int right){
    int pivot = arr[right];
    int i = left-1;
    for(int j=left;j<right;j++){
        if(arr[j]<pivot){
            i++;
            int temp = arr[i];
            arr[i]=arr[j];
            arr[j]=temp;
        }
    }
    i++;
    int temp=arr[i];
    arr[i]=arr[right];
    arr[right]=temp;
    return i;
}
void quickSort(int* arr,int left,int right){
    if(left<right){
        int pi = partition(arr,left,right);
        quickSort(arr,left,pi-1);
        quickSort(arr,pi+1,right);
    }
}
int main(){
    int size;
    printf("Enter the size of array: ");
    scanf("%d",&size);
    int arr[size];
    printf("Enter the elements of array: ");
    for(int i=0;i<size;i++){
         scanf("%d",&arr[i]);
    }
    printf("Before Sort:");
    for(int i=0;i<size;i++){
        printf("%d ",arr[i]);
    }
    printf("\n");
    quickSort(arr,0,size-1);
    printf("\nAfter Sort: ");
    for(int i=0;i<size;i++){
        printf("%d ",arr[i]);
    }
}

File: CDAC/edac_data_structure/src/edac_data_structure/QuickSort.java
package edac_data_structure;

import java.util.Scanner;

public class QuickSort {
	
	static int partition(int[] arr,int start,int end) {
		int pivot = arr[end];
		int i=start-1;
		for(int j=start;j<end;j++) {
			if(arr[j]<pivot) {
				i++;
				int swap = arr[i];
				arr[i]=arr[j];
				arr[j]=swap;
			}
		}
		int swap = arr[end];
		arr[end]=arr[i+1];
		arr[i+1]=swap;
		return i+1;
	}
	
	static void sort(int arr[],int start,int end) {
		if(start<end) {
			int pi = partition(arr,start,end);
			sort(arr,start,pi-1);  //left
			sort(arr,pi+1,end);   //right
		}
	}
	
	public static void main(String[] args) {
		System.out.println("Enter the size of array: ");
		Scanner sc = new Scanner(System.in);
		int size=sc.nextInt();
		int arr[]=new int[size];
		System.out.println("Enter the elements of the array: ");
		for(int i=0;i<arr.length;i++) {
			arr[i]=sc.nextInt();
		}
		sort(arr,0,arr.length-1);   //start=0,end=length-1
		for(int i=0;i<arr.length;i++) {
			System.out.print(arr[i]+" ");
		}
		System.out.println();
		sc.close();
	}

}
